mod-source-record-manager-server/src/main/java/org/folio/services/afterprocessing/AdditionalFieldsUtil.java

```diff
@@ -175,60 +171,65 @@ public static boolean addFieldToMarcRecord(Record record, String field, char sub
+      var parsedContent = objectMapper.readTree(systemOrderContent);
       var fieldsArrayNode = (ArrayNode) parsedContent.path(FIELDS);
 
-      var jsonNodesByTag = groupNodesByTag(fieldsArrayNode);
-      var sourceFields = getSourceFields(sourceContent);
-      var rearrangedArray = objectMapper.createArrayNode();
+      var nodes = toNodeList(fieldsArrayNode);
+      var sourceOrderTags = getSourceFields(sourceOrderContent);
+      var reorderedFields = objectMapper.createArrayNode();
 
-      var nodes001 = jsonNodesByTag.get(TAG_001);
-      if (nodes001 != null && !nodes001.isEmpty()) {
-        rearrangedArray.addAll(nodes001);
-        jsonNodesByTag.remove(TAG_001);
+      var node001 = removeAndGetNodeByTag(nodes, TAG_001);
+      if (node001 != null && !node001.isEmpty()) {
+        reorderedFields.add(node001);
       }
 
-      var nodes005 = jsonNodesByTag.get(TAG_005);
-      if (nodes005 != null && !nodes005.isEmpty()) {
-        rearrangedArray.addAll(nodes005);
-        jsonNodesByTag.remove(TAG_005);
+      var node005 = removeAndGetNodeByTag(nodes, TAG_005);
+      if (node005 != null && !node005.isEmpty()) {
+        reorderedFields.add(node005);
       }
 
-      for (String tag : sourceFields) {
-        Queue<JsonNode> nodes = jsonNodesByTag.get(tag);
-        if (nodes != null && !nodes.isEmpty()) {
-          rearrangedArray.addAll(nodes);
-          jsonNodesByTag.remove(tag);
+      for (String tag : sourceOrderTags) {
+        var node = removeAndGetNodeByTag(nodes, tag);
+        if (node != null && !node.isEmpty()) {
+          reorderedFields.add(node);
         }
-
       }
 
-      jsonNodesByTag.values().forEach(rearrangedArray::addAll);
+      reorderedFields.addAll(nodes);
 
-      ((ObjectNode) parsedContent).set(FIELDS, rearrangedArray);
+      ((ObjectNode) parsedContent).set(FIELDS, reorderedFields);
       return parsedContent.toString();
     } catch (Exception e) {
       LOGGER.error("An error occurred while reordering Marc record fields: {}", e.getMessage(), e);
-      return targetContent;
+      return systemOrderContent;
+    }
+  }
+
+  private static List<JsonNode> toNodeList(ArrayNode fieldsArrayNode) {
+    var nodes = new LinkedList<JsonNode>();
+    for (var node : fieldsArrayNode) {
+      nodes.add(node);
     }
+    return nodes;
   }
 
-  private static Map<String, Queue<JsonNode>> groupNodesByTag(ArrayNode fieldsArrayNode) {
-    var jsonNodesByTag = new LinkedHashMap<String, Queue<JsonNode>>();
-    for (JsonNode node : fieldsArrayNode) {
-      var tag = getTagFromNode(node);
-      jsonNodesByTag.putIfAbsent(tag, new LinkedList<>());
-      jsonNodesByTag.get(tag).add(node);
+  private static JsonNode removeAndGetNodeByTag(List<JsonNode> nodes, String tag) {
+    for (int i = 0; i < nodes.size(); i++) {
+      var nodeTag = getTagFromNode(nodes.get(i));
+      if (nodeTag.equals(tag)) {
+        return nodes.remove(i);
```

Comment:
is it safe to remove in this way while doing traversal?

Reply:
I think better to achieve this either with ```Iterator``` or ```List:: removeIf```

Reply:
I think it's safe because we stop traversing on removal